scm_v3c/teensy_uC_programmer/teensy_uC_programmer.ino

```diff
@@ -0,0 +1,1773 @@
+  digitalWrite(aSCANIN, LOW);
+
+  // Pin for optical TX
+  pinMode(optical_out, OUTPUT);
+  digitalWrite(optical_out, LOW);
+
+}
+
+
+// Runs repeatedly
+// Monitors uart for commands and then makes function call
+void loop() {
+  // Do nothing until a '\n' terminated string is received
+  if (stringComplete) {
+
+    if (inputString == "transfersram\n") {
+      transfer_sram();
+    }
+
+    else if (inputString == "insertcrc\n") {
+      insert_crc();
+    }
+
+    else if (inputString == "transfersram4b5b\n") {
+      transfer_sram_4b5b();
+    }
+
+    else if (inputString == "transfersdigdata\n") {
+      transfer_dig_data();
+    }
+
+    else if (inputString == "digdata\n") {
+      output_digital_data();
+    }
+
+    else if (inputString == "transmit_chips\n") {
+      transmit_chips();
+    }
+
+    else if (inputString == "boot3wb\n") {
+      bootload_3wb();
+    }
+
+    else if (inputString == "boot3wb_debug\n") {
+      bootload_3wb_debug();
+    }
+
+    else if (inputString == "boot3wb4b5b\n") {
+      bootload_3wb_4b5b();
+    }
+
+    else if (inputString == "optdata\n") {
+      optical_data_transfer(p1, p2, p3, p4);
+    }
+
+    else if (inputString == "bootopt4b5b\n") {
+      bootload_opt_4b5b(p1, p2, p3, p4);
+    }
+
+    else if (inputString == "bootopt4b5bnorst\n") {
+      bootload_opt_4b5b_no_reset(p1, p2, p3, p4);
+    }
+
+    else if (inputString == "configopt\n") {
+      optical_config();
+    }
+
+    else if (inputString == "opti_cal\n") {
+      opti_cal();
+    }
+
+    else if (inputString == "3wb_cal\n") {
+      cal_3wb();
+    }
+
+    else if (inputString == "encode4b5b\n") {
+      encode_4b5b();
+    }
+
+    else if (inputString == "ascwrite\n") {
+      asc_write();
+    }
+
+    else if (inputString == "ascload\n") {
+      asc_load();
+    }
+
+    else if (inputString == "ascread\n") {
+      asc_read();
+    }
+
+    else if (inputString == "toggle3wbclk\n") {
+      toggle_3wb_enable();
+    }
+
+    else if (inputString == "dscread\n") {
+      dsc_read();
+    }
+    
+    else if (inputString == "dscdebug\n") {
+      dsc_debug();
+    }
+    
+    else if (inputString == "stepclk\n") {
+      step_clock();
+    }
+
+    else if (inputString == "clockon\n") {
+      clock_on();
+    }
+
+    else if (inputString == "clockoff\n") {
+      clock_off();
+    }
+
+    else if (inputString == "togglehardreset\n") {
+      togglehardreset();
+    }
+
+    // Reset to listen for a new '\n' terminated string over serial
+    inputString = "";
+    stringComplete = false;
+  }
+}
+
+unsigned int parse_bindata_payload(char* payload, int payload_size, unsigned int sram_write_head) {
+  char hexh[3] = {0}; // Add null terminator
+  char* endptr;
+  byte byte_value;
+  // Process the hex data
+  for (int i = 0; i < payload_size; i++) {
+    hexh[0] = payload[i * 2];
+    hexh[1] = payload[i * 2 + 1];
+    hexh[2] = '\0';
+
+    byte_value = (byte)strtoul(hexh, &endptr, 16);
```

Comment:
&endptr can be NULL too

Reply:
It frequently was inexplicably. I used to do NULL checks but it would be randomly NULL even when the payload was deciphered normally.